beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,185 @@
+use safe_arith::SafeArith;
+use slog::{debug, Logger};
+use state_processing::{
+    common::altair::BaseRewardPerIncrement,
+    per_epoch_processing::altair::{participation_cache, rewards_and_penalties::get_flag_weight},
+};
+use std::{collections::HashMap, sync::Arc};
+use store::consts::altair::PARTICIPATION_FLAG_WEIGHTS;
+use types::consts::altair::WEIGHT_DENOMINATOR;
+
+use types::{Epoch, EthSpec};
+
+use crate::{beacon_chain, BeaconChainError};
+
+impl<T: BeaconChainTypes> BeaconChain<T> {
+    pub fn compute_attestation_rewards(
+        &self,
+        epoch: Epoch,
+        chain: Arc<BeaconChain<T>>,
+        mut validators: Vec<usize>,
+        log: Logger,
+    ) -> Result<StandardAttestationRewards, BeaconChainError> {
+        debug!(log, "computing attestation rewards"; "epoch" => epoch, "validator_count" => validators.len());
+
+        //--- Get state ---//
+        let spec = &chain.spec;
+
+        let execution_optimistic = chain.is_optimistic_or_invalid_head()?;
+
+        let state_slot = (epoch + 1).end_slot(T::EthSpec::slots_per_epoch());
+
+        let state_root = chain.state_root_at_slot(state_slot)?;
+
+        let state = chain.get_state(&state_root.unwrap(), Some(state_slot))?;
+
+        //--- Calculate ideal_rewards ---//
+        let participation_cache = ParticipationCache::new(state.as_ref().unwrap(), spec)?;
+
+        let previous_epoch = state.as_ref().unwrap().previous_epoch();
+
+        let mut ideal_rewards_hashmap = HashMap::new();
+
+        for flag_index in PARTICIPATION_FLAG_WEIGHTS {
+            let weight = get_flag_weight(flag_index.try_into().unwrap()).unwrap_or_default();
+
+            let unslashed_participating_indices = participation_cache
+                .get_unslashed_participating_indices(
+                    flag_index.try_into().unwrap(),
+                    previous_epoch,
+                )?;
+
+            let unslashed_participating_balance =
+                unslashed_participating_indices.total_balance().unwrap();
+
+            let unslashed_participating_increments =
+                unslashed_participating_balance.safe_div(spec.effective_balance_increment)?;
+
+            let total_active_balance = participation_cache.current_epoch_total_active_balance();
+
+            let active_increments =
+                total_active_balance.safe_div(spec.effective_balance_increment)?;
+
+            let base_reward_per_increment =
+                BaseRewardPerIncrement::new(total_active_balance, spec)?;
+
+            for effective_balance_eth in 0..=32 {
+                let base_reward =
+                    effective_balance_eth.safe_mul(base_reward_per_increment.as_u64())?;
+
+                let reward_numerator = base_reward
+                    .safe_mul(weight)
+                    .and_then(|reward_numerator| {
+                        reward_numerator.safe_mul(unslashed_participating_increments)
+                    })
+                    .unwrap();
```

Comment:
This `and_then` and `unwrap` can be replaced by `?`

```suggestion
                let reward_numerator = base_reward
                    .safe_mul(weight)?
                    .safe_mul(unslashed_participating_increments)?;
```